---
title: "MULTI-seq Barcode Test"
output: html_notebook
---

```{r}
library(Seurat)
library(patchwork)
library(stringr)
library(ggplot2)
library(dplyr)
library(ggplotify)
library(pheatmap)
```

## Data 

```{r}
ref_genome = read.csv("gem_classification.csv")
data = readRDS("FU180-MULTI-seq_demultiplex.rds")
table(data$HTO_classification)
```

```{r}
rownames(ref_genome) = ref_genome$barcode
ref_genome = ref_genome[rownames(data@meta.data),]

data[["genome"]] = ref_genome$call

table(data$genome)
```

```{r}
GRCh38.features = rownames(data)[str_detect(rownames(data), "^GRCh38.")]
EquCab3.features = rownames(data)[str_detect(rownames(data), "^EquCab3.")]
```

```{r, echo=T, results='hide'}
data[["percent.mt"]] <- PercentageFeatureSet(data, pattern = ".MT-")
```

## Filter Data 

```{r}
Idents(data) = "genome"
VlnPlot(data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```
```{r}
data = subset(data, subset = percent.mt < 6 & nFeature_RNA > 700 & nFeature_RNA < 6000)
```

```{r}
VlnPlot(data, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


```{r, echo=T, results='hide'}
data <- NormalizeData(data, normalization.method = "LogNormalize", scale.factor = 10000)
data <- FindVariableFeatures(data, selection.method = "vst", nfeatures = 2000)
data <- ScaleData(data, features = rownames(data))
data <- RunPCA(data, features = VariableFeatures(object = data))
ElbowPlot(data)
```

```{r}
data <- FindNeighbors(data, dims = 1:10)
data <- FindClusters(data, resolution = 0.05)
data <- RunUMAP(data, dims = 1:10)
```

```{r}
#saveRDS(data, "FU180-MULTI-seq_demultiplex_prepoc.rds")
data = readRDS("FU180-MULTI-seq_demultiplex_prepoc.rds")
```

```{r}
Idents(data) = "HTO_classification"
p1 = DimPlot(data, reduction = "umap", pt.size = 1) +
  scale_color_discrete(name = "Barcode", 
                       labels = c("Negative", "Multiplet", "Barcode 2", "Barcode 1")) +
  ggtitle("") +
   theme(plot.title=element_text(size=30), 
         legend.text=element_text(size = 17),
         legend.title=element_text(size = 17))

Idents(data) = "genome"
p2 = DimPlot(data, reduction = "umap", pt.size = 1) + 
  scale_color_discrete(name = "Gene Expression") +
  ggtitle("") +
   theme(plot.title=element_text(size=30), 
         legend.text=element_text(size = 17),
         legend.title=element_text(size = 17))

Idents(data) = "seurat_clusters"
p3 = DimPlot(data, reduction = "umap", label = T, label.size = 10, pt.size = 1) + 
  scale_color_discrete(name = "Cluster") +
  ggtitle("") +
   theme(plot.title=element_text(size=30), 
         legend.text=element_text(size = 17),
         legend.title=element_text(size = 17))

Idents(data) = "HTO_maxID"
p4 = DimPlot(data, reduction = "umap", label = T, label.size = 10, pt.size = 1) + 
  scale_color_discrete(name = "Max Barcode") +
  ggtitle("") +
   theme(plot.title=element_text(size=30), 
         legend.text=element_text(size = 17),
         legend.title=element_text(size = 20))

p1
p2
p3
p4
```

```{r}
Idents(data) = "seurat_clusters"
markers <- FindAllMarkers(data, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
```

```{r}
#saveRDS(markers, "markers_MULTI-seq.rds")
markers = readRDS("markers_MULTI-seq.rds")
```

```{r}
top_markers = markers %>% group_by(cluster) %>% top_n(n = 100, wt = avg_log2FC)
top_markers
```

```{r}
library(RColorBrewer)

avg.expression = AverageExpression(data, features = top_markers$gene, group.by = "seurat_clusters", assays = "RNA", slot = "scale.data")$RNA
avg.expression = avg.expression[,order(colnames(avg.expression), decreasing = T)]
col = ifelse(str_detect(rownames(avg.expression), "EquCab3"), "royalblue3", "red3")
col[str_detect(rownames(avg.expression), "MT-")] = "green3"
  
heatmap(t(avg.expression), 
        Rowv = NA, 
        Colv = NA,
        ColSideColors = col, 
        labCol = NA, 
        col= colorRampPalette(brewer.pal(8, "Oranges"))(25), 
        main = "D                                                             ")
```

```{r}
avg.expression = AverageExpression(data, features = top_markers$gene, group.by = "seurat_clusters", assays = "RNA", slot = "scale.data")$RNA

col = ifelse(str_detect(rownames(avg.expression), "EquCab3"), "EquCab3", "GRCh38")
col[str_detect(rownames(avg.expression), "MT-")] = "MT"
col = as.data.frame(col, row.names = rownames(avg.expression))

color = list(col = c("EquCab3" = "green4",  "GRCh38" = "orange", "MT" = "red3"))

h1 = as.ggplot(pheatmap(t(avg.expression), 
         cluster_rows = F,
         cluster_cols = F,
         show_colnames = F, 
         annotation_legend = F,
         annotation_names_col = F,
         annotation_col = col,
         annotation_colors = color,
         scale = "row",
         fontsize = 20)) 
```


```{r}
nrow(data@meta.data[data@meta.data$HTO_classification == "cell-type-1" & data@meta.data$seurat_clusters == "3",])
```


```{r}
plot = (p2 | p3) / (p1 | h1)
plot
ggsave(plot = plot, filename = "MULTI-seq_barcodes.jpeg", height = 20, width = 30, units = "cm")
```



```{r}
?heatmap
```

```{r}
table(data@meta.data$seurat_clusters)
```

